### PR TITLE
test(types): add comprehensive unit tests for state.go

### LINKED ISSUE #659: feat: Add Unit Tests for state.go
# **Problem**

The OLake codebase currently has unit tests for:

* `resolver`
* `flattener`
* `catalog`

However, the file `olake/types/state.go` does not have any test coverage.

This file handles critical state management logic for the OLake connector, including:

* **State Management Methods:**
  * `SetType()` - Sets the state type (GLOBAL, STREAM, MIXED)
  * `ResetStreams()` - Resets all stream states
  * `ResetCursor()` - Resets cursor values for a specific stream
  * `isZero()` - Checks if state is empty

* **Global State Operations:**
  * `GetGlobal()` - Retrieves global state with read lock
  * `SetGlobal()` - Sets or updates global state with stream tracking
  * `HasCompletedBackfill()` - Checks if backfill is completed for a stream (supports both GLOBAL and STREAM state types)

* **Cursor Management:**
  * `GetCursor()` - Retrieves cursor value for a stream and key
  * `SetCursor()` - Sets cursor value for a stream (creates new StreamState if needed)

* **Chunk Management:**
  * `GetChunks()` - Retrieves all chunks for a stream
  * `SetChunks()` - Sets chunks for a stream (skips for FULLREFRESH mode)
  * `RemoveChunk()` - Removes a specific chunk and returns remaining count

etc.

Without unit tests, bugs in state management, race conditions with concurrent access, incorrect serialization/deserialization, or edge cases with nil values, empty states, or chunk operations may slip through during development. This is particularly critical as the `State` struct is used extensively across the codebase including:

* Incremental sync operations
* Backfill operations
* CDC (Change Data Capture) processing etc.

---

#### **Proposed Solution**

Add comprehensive unit tests for the logic inside `state.go`.

This will help:

* Increase overall test coverage
* Improve state management reliability across different state types
* Catch bugs earlier during CI, especially around:
  * Concurrent access patterns (read/write locks)
  * State serialization/deserialization edge cases
  * Chunk management operations
  * Cursor value handling
  * Empty state handling
* Make contributions safer when modifying state management logic

For reference on how existing unit tests are structured, you can refer to **PR #273** and the existing `catalog_test.go`, `flatten_test.go` and `resolver_test.go` files.

---

#### **Notes for Contributors**

* Add tests covering most of the scenarios and edge cases which should indicate both pass and fail scenarios.

* Follow the existing test file structure and naming conventions to maintain consistency with other parts (see `catalog_test.go` for reference).

* Use `testify/assert` and `testify/require` for assertions (as seen in existing test files).

### PR DESCRIPTION
**Pull request by rajanarahul93**
# Description

This PR adds a comprehensive suite of unit tests for `types/state.go`, which previously had no test coverage.  
The state manager is central to OLake’s incremental sync, backfill logic, and CDC processes, and verifying its behavior reduces the risk of regressions in cursor handling, chunk management, global state updates, and JSON serialization.

The new tests cover:

- State lifecycle: `isZero`, `SetType`, `ResetStreams`
- Cursor handling: setting, retrieving, and resetting primary and secondary cursors
- Global state operations: `SetGlobal`, `GetGlobal`, and backfill completion logic for both GLOBAL and STREAM state types
- Chunk management: `SetChunks`, `GetChunks`, and `RemoveChunk`
- JSON marshalling/unmarshalling of `State` and `StreamState`
- Behavior of populated vs. unpopulated streams during serialization
- Logging suppression for tests by redirecting `StatePath` to `NUL` or `/dev/null` to avoid filesystem writes

Fixes #659

## Type of change

- [x] Test addition (non-breaking change that adds coverage)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

The following tests were executed:

- Full test suite for the `types` package:
  ```bash
  go test ./types -v

Validation results:

* All existing OLake tests (catalog, resolver, flattener, etc.) passed.
* All new tests passed consistently across platforms.
* No changes were made to production code, ensuring zero functional impact.

# Screenshots or Recordings

N/A This PR contains test-only changes.

## Documentation

* [x] N/A (test changes only; no user-facing or API changes)

## Related PR's (If Any):

N/A

### CODE REVIEW

**Comment by nayanj98** [comment]
@rajanarahul93 Please rebase your PR to staging instead of master

**Comment by nayanj98** [comment]
@rajanarahul93 I have assigned @vaibhav-datazip to review your work.

File: types/state_test.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+	s := newState()
+	assert.True(t, s.isZero(), "new state without streams/global should be zero")
+
+	s.SetType(GlobalType)
+	assert.Equal(t, GlobalType, s.Type, "SetType should set the state type")
+
+	// add a stream then reset
+	cfg := newConfiguredStream("s1", "ns1", "id", SyncMode("incremental"))
+	s.SetCursor(cfg, "id", 123)
+	require.False(t, s.isZero(), "state should not be zero after adding cursor")
```
Comment (vaibhav-datazip) [inline types/state_test.go]: why are we using individual assertion and defining so many functions, why not use the existing pattern of single function with many test cases ?

File: types/state_test.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+func newConfiguredStream(name, namespace, cursor string, mode SyncMode) *ConfiguredStream {
+	s := NewStream(name, namespace, nil)
+	s.CursorField = cursor
+	s.SyncMode = mode
+	return s.Wrap(0)
+}
+
+func TestIsZeroAndSetType_ResetStreams(t *testing.T) {
+	s := newState()
+	assert.True(t, s.isZero(), "new state without streams/global should be zero")
```
Comment (vaibhav-datazip) [inline types/state_test.go]: adding more edge cases would be better. like in iszero function there checks for global nil and stream Len check. 

similar for other functions as well.

File: types/state_test.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+func newConfiguredStream(name, namespace, cursor string, mode SyncMode) *ConfiguredStream {
+	s := NewStream(name, namespace, nil)
+	s.CursorField = cursor
+	s.SyncMode = mode
+	return s.Wrap(0)
+}
+
+func TestIsZeroAndSetType_ResetStreams(t *testing.T) {
```
Comment (vaibhav-datazip) [inline types/state_test.go]: function name should follow similar pattern as we have done in other test files, the name contains individual function name as mentioned in original file

**Comment by rajanarahul93** [comment]
Good points, thanks for the feedback.

I’ll refactor this to follow the existing table-driven test pattern used across the repo, split concerns into separate test functions per method, and add explicit edge cases for `isZero` (global state vs stream length).

I’ll also update the test names to follow the `<Struct>_<Method>` convention used in other `types` tests.

File: types/state_test.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+func TestState_MarshalJSON_PopulatedStreamsOnly(t *testing.T) {
+	s := newState()
+	cfg1 := newConfiguredStream("a", "n", "id", SyncMode("incremental"))
+	cfg2 := newConfiguredStream("b", "n", "id", SyncMode("incremental"))
+
+	// stream a: set cursor -> holds value
+	s.SetCursor(cfg1, "id", 1)
+	// stream b: create streamstate but not hold any value
```
Comment (vaibhav-datazip) [inline types/state_test.go]: even for other functions as well you haven't followed the structure of our code, for mashaljson and unmarshal json you are still using separate function for each test case.
check other functions as well which I haven't mentioned

**Comment by vaibhav-datazip** [comment]
@rajanarahul93 please re-request review or drop a comment  whenever this pr is again ready for review

**Comment by vaibhav-datazip** [comment]
@rajanarahul93 Also, to understand how the state.go file is used by the rest of the codebase, please run the sync once and observe the edge cases that get triggered. This will help you reason about the edge cases more effectively.

You can refer to the [Olake commands and flags documentation](https://olake.io/docs/community/commands-and-flags/) to understand the different places where the state file can be passed or manipulated.

Please attach a short video of you running the sync here as well. This will help me understand how familiar you are with the flow, and I can then review any edge cases that might have been missed and let you know.

You can also [join the Slack](https://join.slack.com/t/getolake/shared_invite/zt-2uyphqf69-KQxih9Gwd4GCQRD_XFcuyw) and stay active there in case you miss messages on GitHub.

**Comment by vaibhav-datazip** [comment]
Hi @rajanarahul93 , as there was no response from your side for a week, I am closing this PR.

**Comment by rajanarahul93** [comment]
Sorry for that, i was recently kind of busy so couldn't make the changes. Will pick other issues any other time!!
